flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java

```diff
@@ -96,332 +222,274 @@ public void initializeState(StateInitializationContext context) throws Exception
-		this.reader = new SplitReader<>(format, serializer, readerContext, checkpointLock, restoredReaderState);
-		this.restoredReaderState = null;
-		this.reader.start();
+
+		this.sourceContext = StreamSourceContexts.getSourceContext(
+				getOperatorConfig().getTimeCharacteristic(),
+				getProcessingTimeService(),
+				new Object(), // no actual locking needed
+				getContainingTask().getStreamStatusMaintainer(),
+				output,
+				getRuntimeContext().getExecutionConfig().getAutoWatermarkInterval(),
+				-1);
+
+		this.reusedRecord = serializer.createInstance();
+		this.completedSplitsCounter = getMetricGroup().counter("numSplitsProcessed");
+		this.executor = getContainingTask().getMailboxExecutorFactory().createExecutor(getOperatorConfig().getChainIndex());
+		this.splits = this.splits == null ? new PriorityQueue<>() : this.splits;
+
+		if (!splits.isEmpty()) {
+			enqueueMail();
+		}
 	}
 
 	@Override
 	public void processElement(StreamRecord<TimestampedFileInputSplit> element) throws Exception {
-		reader.addSplit(element.getValue());
+		Preconditions.checkState(state.isAcceptingSplits());
+		splits.offer(element.getValue());
+		if (state == ReaderState.IDLE) {
+			enqueueMail();
+		}
 	}
 
-	@Override
-	public void processWatermark(Watermark mark) throws Exception {
-		// we do nothing because we emit our own watermarks if needed.
+	private void enqueueMail() {
+		Preconditions.checkState(!state.isTerminal(), "can't enqueue mail in terminal state %s", state);
+		executor.execute(this, "ContinuousFileReaderOperator");
+		if (state == ReaderState.IDLE) {
+			switchState(ReaderState.OPENING);
+		}
 	}
 
 	@Override
-	public void dispose() throws Exception {
-		super.dispose();
-
-		// first try to cancel it properly and
-		// give it some time until it finishes
-		reader.cancel();
+	public void run() { // enqueued as Runnable into the mailbox (executor)
 		try {
-			reader.join(200);
-		} catch (InterruptedException e) {
-			// we can ignore this
-		}
-
-		// if the above did not work, then interrupt the thread repeatedly
-		while (reader.isAlive()) {
-
-			StringBuilder bld = new StringBuilder();
-			StackTraceElement[] stack = reader.getStackTrace();
-			for (StackTraceElement e : stack) {
-				bld.append(e).append('\n');
-			}
-			LOG.warn("The reader is stuck in method:\n {}", bld.toString());
-
-			reader.interrupt();
-			try {
-				reader.join(50);
-			} catch (InterruptedException e) {
-				// we can ignore this
-			}
+			processRecord();
+		} catch (Exception e) {
+			switchState(ReaderState.CLOSED);
+			cleanUp();
+			throw new FlinkRuntimeException("unable to process record from file split", e);
 		}
-		reader = null;
-		readerContext = null;
-		restoredReaderState = null;
-		format = null;
-		serializer = null;
 	}
 
-	@Override
-	public void close() throws Exception {
-		super.close();
+	private void processRecord() throws IOException {
+		if (!state.prepareToProcessRecord(this)) {
+			return;
```

Comment:
How can that happen? Should that result in some kind of exception?

Reply:
In `CLOSING` state when there is no more data
it enqueues one more mail and transitions to `CLOSED`.

Additional mail is needed to unblock `yield()` call in `close()`.